app/src/main/java/org/piwigo/bg/UploadService.java

```diff
@@ -143,12 +151,74 @@ protected void onUploadStarted(ImageUploadQueue<UploadAction> imageUploadQueue)
             Log.e("UploadService", "IOException", e.getCause());
             content = new byte[0];
         }
-        filePart = MultipartBody.Part.createFormData("file", uploadAction.getFileName(), RequestBody.create(MediaType.parse("image/*"), content));
+
+        byte[][] splittedContent = splitArray(content, chunkSize);
+        for (int i = 0; i < splittedContent.length; i++) {
+            fileParts[i] = MultipartBody.Part.createFormData("file", uploadAction.getFileName(), RequestBody.create(MediaType.parse("image/*"), splittedContent[i]));
+        }
         restService = restServiceFactory.createForAccount(userManager.getActiveAccount().getValue());
         requestBodies = createUploadRequest(uploadAction.getFileName(), getPhotoName(uploadAction.getFileName()), userManager.getActiveAccount().getValue());
 
+        //Promise building
+        promise.setFileParts(fileParts);
+        promise.setRequestBodies(requestBodies);
+        promise.setCatId(uploadAction.getUploadData().getCategoryId());
+        promise.setCurrentChunk(0);
+        promise.setTotalChunks(expectedChunks);
         if (uploadAction.getUploadData().getCategoryId() > 0)
-            callUploadResponse(imageUploadQueue, restService, requestBodies, uploadAction.getUploadData().getCategoryId(), filePart);
+            callChunkedUploadResponse(imageUploadQueue, restService, promise);
+    }
+
+    /**
+     * Pick a chunk size depending of the network mode (if on wifi, 1MB, 256KB for mobile data)
+     * @return chunkSize (1MB, 256KB for mobile data or anything else)
+     */
+    public int getChunkSize()
+    {
+        int networkType = NetworkHelper.INSTANCE.getNetworkType(getApplicationContext());
+
+        if (networkType == ConnectivityManager.TYPE_WIFI) {
```

Comment:
I like the idea of deciding the size based on connection type, still I think it would be wise to either autodetect the junk size (no idea how to find out how the server is configured, but piwigo server knows it, so I believe there is a REST service to read it) or at least provide a configuration option (but this only in case we see no chance to autodetect.

If we know the maximum size we should limit the decision taken here by the max server allowed size.

Reply:
In https://github.com/Piwigo/Piwigo-Android/pull/214/commits/32a76595cf1b40cecafbc36be2720e8544eb47c2 I'm using the size given by Piwigo, used in the web upload form.